crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1456,47 +1458,79 @@ impl Client {
                 &[MatrixVersion::V1_0],
                 Default::default(),
             )
-            .await?
-            .known_versions()
-            .collect();
+            .await?;
 
-        if server_versions.is_empty() {
-            Ok(vec![MatrixVersion::V1_0].into())
-        } else {
-            Ok(server_versions)
+        // Fill both unstable features and server versions at once.
+        let mut versions = resp.known_versions().collect::<Vec<_>>();
+        if versions.is_empty() {
+            versions.push(MatrixVersion::V1_0);
         }
+
+        let unstable_features = resp.unstable_features;
+
+        {
+            // Attempt to cache the result.
+            let encoded = ServerCapabilities::new(&versions, unstable_features.clone());
+            if let Err(err) = self
+                .store()
+                .set_kv_data(
+                    StateStoreDataKey::ServerCapabilities,
+                    StateStoreDataValue::ServerCapabilities(encoded),
+                )
+                .await
+            {
+                warn!("error when caching server capabilities: {err}");
+            }
+        }
+
+        Ok((versions.into(), unstable_features))
     }
 
-    pub(crate) async fn server_versions(&self) -> HttpResult<&[MatrixVersion]> {
-        let server_versions = self
-            .inner
-            .server_versions
-            .get_or_try_init(|| Box::pin(self.request_server_versions()))
-            .await?;
+    async fn load_or_fetch_server_capabilities(
+        &self,
+    ) -> HttpResult<(Box<[MatrixVersion]>, BTreeMap<String, bool>)> {
+        match self.store().get_kv_data(StateStoreDataKey::ServerCapabilities).await {
+            Ok(Some(stored)) => {
+                if let Some((versions, unstable_features)) =
+                    stored.into_server_capabilities().and_then(|cap| cap.maybe_decode())
+                {
+                    return Ok((versions.into(), unstable_features));
+                }
+            }
+            Ok(None) => {
+                // fallthrough: cache is empty
```

Comment:
It's OK to have a call to `self.fetch_and_cache_server...` twice, I would say that this is more reasonable than having an early return and two empty clauses explaining why they are empty.

Reply:
*narrator's voice* there are in fact not two, but three fallthrough paths, if the `Ok(Some())`'s `if` is not taken. I don't quite see the issue with the comments?

Reply:
Alright.